extensions/jackson/runtime/src/main/java/io/quarkus/jackson/runtime/HybridJacksonPool.java

```diff
@@ -0,0 +1,141 @@
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Predicate;
+
+public class HybridJacksonPool implements BufferRecyclerPool {
+
+    static final BufferRecyclerPool INSTANCE = new HybridJacksonPool();
+
+    private HybridJacksonPool() { }
+
+    private static final Predicate<Thread> isVirtual = findIsVirtual();
+
+    private static Predicate<Thread> findIsVirtual() {
+        try {
+            MethodHandle virtualMh = MethodHandles.publicLookup().findVirtual(Thread.class, "isVirtual", MethodType.methodType(boolean.class));
+            return t -> {
+                try {
+                    return (boolean) virtualMh.invokeExact(t);
+                } catch (Throwable e) {
+                    throw new RuntimeException(e);
+                }
+            };
+        } catch (Exception e) {
+            return t -> false;
+        }
+    }
+
+    private final BufferRecyclerPool nativePool = BufferRecyclerPool.threadLocalPool();
+
+    static class VirtualPoolHolder {
+        // Lazy on-demand initialization
+        private static final BufferRecyclerPool virtualPool = new StripedJCToolsPool(4);
+    }
+
+    @Override
+    public BufferRecycler acquireBufferRecycler() {
+        return isVirtual.test(Thread.currentThread()) ?
+                VirtualPoolHolder.virtualPool.acquireBufferRecycler() :
+                nativePool.acquireBufferRecycler();
+    }
+
+    @Override
+    public void releaseBufferRecycler(BufferRecycler bufferRecycler) {
+        if (bufferRecycler instanceof VThreadBufferRecycler) {
+            // if it is a PooledBufferRecycler it has been acquired by a virtual thread, so it has to be release to the same pool
+            VirtualPoolHolder.virtualPool.releaseBufferRecycler(bufferRecycler);
+        }
+        // the native thread pool is based on ThreadLocal, so it doesn't have anything to do on release
+    }
+
+    private static class StripedJCToolsPool implements BufferRecyclerPool {
+
+        private static final long PROBE = getProbeOffset();
+
+        private final int mask;
+
+        private final MpmcUnboundedXaddArrayQueue<BufferRecycler>[] queues;
+
+        public StripedJCToolsPool(int stripesCount) {
+            if (stripesCount <= 0) {
+                throw new IllegalArgumentException("Expecting a stripesCount that is larger than 0");
+            }
+
+            int size = Pow2.roundToPowerOfTwo(stripesCount);
+            mask = (size - 1);
+
+            this.queues = new MpmcUnboundedXaddArrayQueue[size];
+            for (int i = 0; i < size; i++) {
+                this.queues[i] = new MpmcUnboundedXaddArrayQueue<>(128);
+            }
+        }
+
+        private static long getProbeOffset() {
+            try {
+                return UnsafeAccess.UNSAFE.objectFieldOffset(Thread.class.getDeclaredField("threadLocalRandomProbe"));
```

Comment:
@cescoffier there's anything to do here considering that we use the UnsafeAccess from JCTools AND probably native image won't be happy about it